lib/peep/buckets/custom.ex

```diff
@@ -7,62 +7,76 @@ defmodule Peep.Buckets.Custom do
   """
 
   defmacro __using__(opts) do
-    buckets =
-      Keyword.fetch!(opts, :buckets)
-      |> :lists.usort()
+    quote bind_quoted: [buckets: Keyword.fetch!(opts, :buckets)] do
+      @behaviour Peep.Buckets
 
-    unless Enum.all?(buckets, &is_number/1) do
-      raise ArgumentError, "expected buckets to be a list of numbers, got: #{buckets}"
-    end
+      require Peep.Buckets.Custom
 
-    number_of_buckets = length(buckets)
+      buckets = :lists.usort(buckets)
+      len = length(buckets)
 
-    quote do
-      @behaviour Peep.Buckets
+      unless Enum.all?(buckets, &is_number/1) do
+        raise ArgumentError, "expected buckets to be a list of numbers, got: #{buckets}"
+      end
 
       @impl true
       def config(_), do: %{}
 
       @impl true
-      def number_of_buckets(_), do: unquote(number_of_buckets)
+      def number_of_buckets(_), do: unquote(len)
 
       @impl true
-      unquote(bucket_for_ast(buckets))
-
-      @impl true
-      unquote(upper_bound_ast(buckets))
-    end
-  end
-
-  ## Bucket binary search
 
-  defp bucket_for_ast(buckets) do
-    int_buckets = int_buckets(buckets, nil, 0)
+      int_buckets =
+        buckets
+        |> Enum.map(&ceil/1)
+        |> Enum.with_index()
+        |> Enum.dedup_by(&elem(&1, 0))
+
+      float_buckets =
+        buckets
+        |> Enum.map(&(&1 * 1.0))
+        |> Enum.with_index()
+
+      if len <= 32 do
+        # For small bucket length it seems that the linear search is faster than
+        # binary search
+        for {top, idx} <- int_buckets do
+          def bucket_for(val, _) when is_integer(val) and val < unquote(top), do: unquote(idx)
+        end
 
-    float_buckets =
-      buckets
-      |> Enum.map(&(&1 * 1.0))
-      |> Enum.with_index()
+        for {top, idx} <- float_buckets do
+          def bucket_for(val, _) when is_float(val) and val < unquote(top), do: unquote(idx)
+        end
 
-    variable = Macro.var(:x, nil)
+        def bucket_for(_, _), do: unquote(length(buckets))
```

Comment:
Here you could reuse `len`, right?
```suggestion
        def bucket_for(_, _), do: unquote(len)
```